README.md
[![Build](https://img.shields.io/github/actions/workflow/status/php-perfect/aviation-calculator/rust.yml?branch=master "Build")](https://github.com/php-perfect/aviation-calculator/actions/workflows/rust.yml)
[![License](https://img.shields.io/badge/License-MIT-blue.svg "MIT")](https://opensource.org/license/mit/)
[![GitHub Sponsors](https://img.shields.io/github/sponsors/nico-loeber?label=Sponsors&logo=Github "Sponsors")](https://github.com/sponsors/nico-loeber?frequency=recurring&sponsor=nico-loeber)

# Aviation Calculator

The goal of this library is to provide a collection of useful calculations for aviators to use in other projects.

The functions are grouped by the general topics of navigation, meteorology and utilities, which mainly contain unit
conversion functions. There are also FK9 Mk VI specific performance calculations.

All calculations are based on official sources and use official examples to verify correctness in automated tests.

Feel free to use this library in your projects and to [contribute](CONTRIBUTING.md) to it.

## Usage

You can install this library as a crate using the rust package manager cargo by running `cargo add aviation-calculator`
in your project.

For a complete list of available functions, see the automatically
generated [documentation](https://docs.rs/crate/aviation-calculator/latest/target-redirect/x86_64-unknown-linux-gnu/aviation_calculator/index.html)
for this crate.

### Some Examples

#### Get ICAO Default temperature for your current location

```rust
use aviation_calculator::meteorology::*;

let pressure_altitude: f64 = pressure_altitude_by_qnh(996.0, 113.7);
let icao_default_temperature = icao_temperature(pressure_altitude).unwrap();
```

#### Calculate wind correction angle

```rust
use aviation_calculator::navigation::*;

let my_speed = 110.0;
let wind_speed = 12.0;
let acute_wind_angle = 20.0;

let wca = wind_correction_angle(my_speed, wind_speed, acute_wind_angle);
```

And much more...

## Sources

* [ICAO-Standardatmosphäre (ISA) - Deutscher Wetterdienst](https://www.dwd.de/DE/service/lexikon/begriffe/S/Standardatmosphaere_pdf.pdf?__blob=publicationFile&v=3)
* [Flight Manual FK 9 Mk VI - B & F Technik Vertriebs GmbH Speyer – Germany](http://www.flugservice-speyer.de/Flugservice_Speyer/Manuals_files/FK9%20Mk%20VI%20Rev%204%20English.pdf)
* [Flugsicherheitsmitteilung (fsm) 3/75 - http://www.pilotundrecht.de/](http://www.pilotundrecht.de/TEXTE/DOWNLOAD/FSM/fsm_75_3.pdf)

## Supported by

[<img alt="JetBrains" height="100" src="./img/lsv-z.png" title="LSV Zulu" width="100"/>](https://lsv-zellhausen.de/)
[<img alt="JetBrains" height="100" src="https://resources.jetbrains.com/storage/products/company/brand/logos/jb_beam.png" title="JetBrains" width="100"/>](https://jb.gg/OpenSourceSupport)


CHANGELOG.md
# Changelog

## 0.2.2

* Set sensible maximum and minimum border for temperature and slope

## 0.2.1

* Make GrassSurface properties public
* Rename SurfaceCondition::Dry to Inconspicuous as it can be combined with wet gras surface

## 0.2.0

* Add social files ([README.md](README.md), [CONTRIBUTING.md](CONTRIBUTING.md))
* Add FK9 Mk VI takeoff calculations
* Add a lot of new test cases


tests/integration_test.rs
use aviation_calculator::fk9::*;
use aviation_calculator::meteorology::pressure_altitude_by_qnh;
use aviation_calculator::utils::{feet_to_meter, meter_to_feet, round};

#[test]
fn common_takeoff_calculation_zellhausen() {
    let result: TakeoffResult = calculate_takeoff_distance(Engine::Rotax912Uls, 520.0, 370.73, 21.0, 0.0, Some(GrassSurface {
        wet: true,
        ..GrassSurface::default()
    }), SurfaceCondition::Inconspicuous);
    let takeoff_distances = result.expect("No error is expected for this takeoff calculation!");
    assert_eq!(takeoff_distances, (152.6, 378.6));
}

#[test]
fn common_takeoff_calculation_frankfurt() {
    let result: TakeoffResult = calculate_takeoff_distance(Engine::Rotax912Uls, 520.0, 364.0, 21.0, 0.0, None, SurfaceCondition::Inconspicuous);
    let takeoff_distances = result.expect("No error is expected for this takeoff calculation!");
    assert_eq!(takeoff_distances, (115.52, 286.61));
}

#[test]
fn common_pressure_altitude() {
    let result: f64 = pressure_altitude_by_qnh(996.0, 113.7);
    assert_eq!(result, 258.25);
}

#[test]
fn pressure_altitude_in_feet() {
    let result: f64 = round(meter_to_feet(pressure_altitude_by_qnh(996.0, feet_to_meter(364.0))), 1);
    assert_eq!(result, 838.2);
}

#[test]
fn pressure_altitude_in_feed_check_rounding() {
    let result: f64 = round(meter_to_feet(pressure_altitude_by_qnh(1013.25, feet_to_meter(364.0))), 1);
    assert_eq!(result, 364.0);
}


src/utils.rs
use std::f64::consts::PI;

const FEET: f64 = 0.3048_f64; /* m */

/// # Convert meter to feet
///
/// ## Arguments
///
/// * `meter`: Value in meters
///
/// returns: f64 Value in feet
///
/// # Examples
///
/// ```
/// use aviation_calculator::utils::*;
///
/// let feet = meter_to_feet(50.0);
/// ```
pub fn meter_to_feet(meter: f64) -> f64 {
    meter / FEET
}

/// # Convert feet to meter
///
/// ## Arguments
///
/// * `feet`: Value in feet
///
/// returns: f64 Value in meter
///
/// # Examples
///
/// ```
/// use aviation_calculator::utils::*;
///
/// let meter = feet_to_meter(50.0);
/// ```
pub fn feet_to_meter(feet: f64) -> f64 {
    feet * FEET
}

/// # Convert to Degree
///
/// ## Arguments
///
/// * `value`: Value in radians
///
/// returns: f64 value in degrees
///
/// # Examples
///
/// ```
/// use aviation_calculator::utils::*;
///
/// let degree = to_degree(std::f64::consts::FRAC_PI_2);
/// ```
pub fn to_degree(value: f64) -> f64 {
    180.0_f64 / PI * value
}

/// # Convert to Radian
///
/// ## Arguments
///
/// * `value`: Value in degrees
///
/// returns: f64 Value in radians
///
/// # Examples
///
/// ```
/// use aviation_calculator::utils::*;
///
/// let radians = to_radian(90.0);
/// ```
pub fn to_radian(value: f64) -> f64 {
    PI / 180.0_f64 * value
}

/// # Normalize Degree
///
/// ## Arguments
///
/// * `value`: Any degree value
///
/// returns: f64 Representation of the degree value between 0 and 360
///
/// # Examples
///
/// ```
/// use aviation_calculator::utils::*;
///
/// let degree = normalize_degree(370.0);
/// ```
pub fn normalize_degree(value: f64) -> f64 {
    value % 360.0_f64
}

/// # Round
///
/// ## Arguments
///
/// * `number`: Number to round
/// * `precision`: Precision to round
///
/// returns: f64 Rounded value
///
/// # Examples
///
/// ```
/// use aviation_calculator::utils::*;
///
/// let rounded = round(55.5555, 2);
/// ```
pub fn round(number: f64, precision: u8) -> f64 {
    let base = 10_i32.pow(precision.into()) as f64;
    (number * base).round() / base
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn meter_to_feet_1() {
        let result = meter_to_feet(1.0);
        assert_eq!(result, 3.280839895013123);
    }

    #[test]
    fn meter_to_feet_5() {
        let result = meter_to_feet(5.5);
        assert_eq!(result, 18.04461942257218);
    }

    #[test]
    fn feet_to_meter_1() {
        let result = feet_to_meter(1.0);
        assert_eq!(result, 0.3048);
    }

    #[test]
    fn feet_to_meter_5() {
        let result = feet_to_meter(5.5);
        assert_eq!(result, 1.6764000000000001);
    }

    #[test]
    fn to_degree_1() {
        let result = to_degree(std::f64::consts::FRAC_PI_2);
        assert_eq!(result, 90.0);
    }

    #[test]
    fn to_radian_1() {
        let result = to_radian(90.0);
        assert_eq!(result, std::f64::consts::FRAC_PI_2);
    }

    #[test]
    fn normalize_degree_1() {
        let result = normalize_degree(370.0);
        assert_eq!(result, 10.0);
    }

    #[test]
    fn normalize_degree_2() {
        let result = normalize_degree(360.0);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn round_1() {
        let result = round(55.5555, 2);
        assert_eq!(result, 55.56);
    }
}


src/meterology.rs
// https://www.dwd.de/DE/service/lexikon/begriffe/S/Standardatmosphaere_pdf.pdf;jsessionid=0749F4152999FCCDFB9CD32F32A3B866.live21073?__blob=publicationFile&v=3

use std::error::Error;
use std::fmt;

const ISA_TEMPERATURE: f64 = 288.15_f64; /* K */
const ISA_PRESSURE: f64 = 1013.25_f64; /* hPa */
const TROPOSPHERIC_TEMPERATURE_LAPSE: f64 = 0.0065_f64; /* K m-1 */
const STRATOSPHERIC_TEMPERATURE_LAPSE: f64 = 0.0010_f64; /* K m-1 */
const SPECIFIC_GAS_CONSTANT: f64 = 287.058_f64;
const GRAVITATIONAL_ACCELERATION: f64 = 9.81_f64; /* m/s */
const ICAO_MINIMUM_ELEVATION: f64 = -1_000.0_f64; /* m */
const ICAO_MAXIMUM_ELEVATION: f64 = 80_000.0_f64; /* m */

#[derive(Debug, Clone)]
pub struct UndefinedError(String);

impl Error for UndefinedError {}

impl fmt::Display for UndefinedError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.0)
    }
}

struct AtmosphericLevel {
    base: u32,
    lapse_rate: f64,
    base_temperature: f64,
}

const LEVELS: [&AtmosphericLevel; 4] = [
    // Troposphere
    &AtmosphericLevel {
        base: 0,
        lapse_rate: TROPOSPHERIC_TEMPERATURE_LAPSE,
        base_temperature: 15.0,
    },
    // Tropopause
    &AtmosphericLevel {
        base: 11_000,
        lapse_rate: 0.0,
        base_temperature: -56.5,
    },
    // Stratosphere
    &AtmosphericLevel {
        base: 20_000,
        lapse_rate: STRATOSPHERIC_TEMPERATURE_LAPSE,
        base_temperature: -56.5,
    },
    // Stratosphere
    &AtmosphericLevel {
        base: 32_000,
        lapse_rate: 0.0,
        base_temperature: -44.5,
    },
];

///
///
/// # Arguments
///
/// * `elevation`: Elevation in meters to get the ICAO default temperature for
///
/// returns: Result<f64, UndefinedError>
///
/// # Examples
///
/// ```
/// let temp: f64 = icao_temperature_by_elevation(113.7).unwrap();
/// ```
pub fn icao_temperature_by_elevation(elevation: f64) -> Result<f64, UndefinedError> {
    if elevation < ICAO_MINIMUM_ELEVATION {
        return Err(UndefinedError(
            "The ICAO standard atmosphere below -1 kilometer (-3,280 feet) is not defined.".into(),
        ));
    }

    if elevation > ICAO_MAXIMUM_ELEVATION {
        return Err(UndefinedError(
            "The ICAO standard atmosphere above 80 kilometers (262,500 feet) is not defined."
                .into(),
        ));
    }

    let current_level = atmospheric_level_by_geopotential_altitude(elevation);

    Ok(round(
        current_level.base_temperature
            - (elevation - current_level.base as f64) * current_level.lapse_rate,
        2,
    ))
}

///
///
/// # Arguments
///
/// * `qnh`: QNH for the location given in hPa
/// * `field_elevation`: Field elevation in meters
///
/// returns: f64
///
/// # Examples
///
/// ```
/// let pressure: f64 = pressure_altitude(1021, 113.7);
/// ```
pub fn pressure_altitude(qnh: f64, field_elevation: f64) -> f64 {
    field_elevation
        + ISA_TEMPERATURE / TROPOSPHERIC_TEMPERATURE_LAPSE
        * (1.0_f64
        - (qnh / ISA_PRESSURE).powf(
        SPECIFIC_GAS_CONSTANT * TROPOSPHERIC_TEMPERATURE_LAPSE
            / GRAVITATIONAL_ACCELERATION,
    ))
}

fn atmospheric_level_by_geopotential_altitude<'a>(elevation: f64) -> &'a AtmosphericLevel {
    LEVELS.iter()
        .take_while(|level| elevation >= level.base as f64)
        .last()
        .unwrap_or(LEVELS.first().unwrap())
}

fn round(x: f64, decimals: u8) -> f64 {
    let y = 10_i32.pow(decimals.into()) as f64;
    (x * y).round() / y
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pressure_altitude_example_1() {
        let result = pressure_altitude(1021.0, 113.0);
        assert_eq!(result, 48.70703054690429);
    }

    #[test]
    fn pressure_altitude_example_2() {
        let result = pressure_altitude(1013.25, 113.0);
        assert_eq!(result, 113.0);
    }

    #[test]
    fn pressure_altitude_example_3() {
        let result = pressure_altitude(1021.0, 113.0);
        assert_eq!(result, 48.70703054690429);
    }

    #[test]
    fn isa_temperature_out_or_range_negative() {
        let result = icao_temperature_by_elevation(-1000.01);
        assert!(result.is_err());
    }

    #[test]
    fn isa_temperature_negative_1000() {
        let result = icao_temperature_by_elevation(-1000.0);
        assert_eq!(result.unwrap(), 21.5_f64);
    }

    #[test]
    fn isa_temperature_0() {
        let result = icao_temperature_by_elevation(0.0);
        assert_eq!(result.unwrap(), 15.0_f64);
    }

    #[test]
    fn isa_temperature_113() {
        let result = icao_temperature_by_elevation(113.7);
        assert_eq!(result.unwrap(), 14.26_f64);
    }

    #[test]
    fn isa_temperature_1000() {
        let result = icao_temperature_by_elevation(1000.0);
        assert_eq!(result.unwrap(), 8.5_f64);
    }

    #[test]
    fn isa_temperature_2000() {
        let result = icao_temperature_by_elevation(2000.0);
        assert_eq!(result.unwrap(), 2.0_f64);
    }

    #[test]
    fn isa_temperature_3000() {
        let result = icao_temperature_by_elevation(3000.0);
        assert_eq!(result.unwrap(), -4.5_f64);
    }

    #[test]
    fn isa_temperature_5000() {
        let result = icao_temperature_by_elevation(5000.0);
        assert_eq!(result.unwrap(), -17.5_f64);
    }

    #[test]
    fn isa_temperature_7000() {
        let result = icao_temperature_by_elevation(7000.0);
        assert_eq!(result.unwrap(), -30.5_f64);
    }

    #[test]
    fn isa_temperature_9000() {
        let result = icao_temperature_by_elevation(9_000.0);
        assert_eq!(result.unwrap(), -43.5_f64);
    }

    #[test]
    fn isa_temperature_11000() {
        let result = icao_temperature_by_elevation(11_000.0);
        assert_eq!(result.unwrap(), -56.5_f64);
    }

    #[test]
    fn isa_temperature_13000() {
        let result = icao_temperature_by_elevation(13_000.0);
        assert_eq!(result.unwrap(), -56.5_f64);
    }

    #[test]
    fn isa_temperature_20000() {
        let result = icao_temperature_by_elevation(20_000.0);
        assert_eq!(result.unwrap(), -56.5_f64);
    }

    #[test]
    fn isa_temperature_25000() {
        let result = icao_temperature_by_elevation(25_000.0);
        assert_eq!(result.unwrap(), -61.5_f64);
    }

    #[test]
    fn isa_temperature_32000() {
        let result = icao_temperature_by_elevation(32_000.0);
        assert_eq!(result.unwrap(), -44.5_f64);
    }

    #[test]
    fn isa_temperature_80000() {
        let result = icao_temperature_by_elevation(80_000.0);
        assert_eq!(result.unwrap(), -44.5_f64);
    }

    #[test]
    fn isa_temperature_out_of_range_positive() {
        let result = icao_temperature_by_elevation(80_000.01_f64);
        assert!(result.is_err());
    }
}


src/navigation.rs
use crate::utils::*;

/// # Calculate Ground Speed (GS)
///
/// ## Arguments
///
/// * `course`: Course in degrees
/// * `tas`: True Air Speed (TAS) in any unit
/// * `wd`: Wind Direction (WD) in degrees
/// * `ws`: Wind Speed (WS) in the same unit as tas
///
/// returns: f64 GS in the same unit as TAS is provided
///
/// # Examples
///
/// ```
/// use aviation_calculator::navigation::*;
///
/// let gs = ground_speed(140.0, 110.0, 90.0, 12.0);
/// ```
pub fn ground_speed(course: f64, tas: f64, wd: f64, ws: f64) -> f64 {
    if ws == 0.0 {
        return tas;
    }

    let crs = to_radian(normalize_degree(course));
    let wind_dir = to_radian(normalize_degree(wd));
    let swc = (ws / tas) * (wind_dir - crs).sin();

    round(tas * (1.0 - swc.powi(2)).sqrt() - (ws * (wind_dir - crs).cos()), 2)
}

/// # Calculate Wind Correction Angle (WCA)
///
/// ## Arguments
///
/// * `tas`: True Air Speed (TAS) in any unit
/// * `ws`: Wind Speed (WS) in the same unit as tas
/// * `awa`: Acute Wind Angle (AWA) in degrees
///
/// returns: f64 Wind Correction Angle (WCA) in degrees
///
/// # Examples
///
/// ```
/// use aviation_calculator::navigation::*;
///
/// let wca = wind_correction_angle(110.0, 12.0, 20.0);
/// ```
pub fn wind_correction_angle(tas: f64, ws: f64, awa: f64) -> f64 {
    if awa == 0.0 || awa == 180.0 || ws == 0.0 {
        return 0.0;
    }

    round(to_degree((ws / tas * to_radian(normalize_degree(awa)).sin()).asin()), 2)
}

/// # Calculate Heading
///
/// ## Arguments
///
/// * `dc`: Desired Course (DC)
/// * `tas`: True Air Speed (TAS)
/// * `wd`: Wind Direction (WD)
/// * `ws`: Wind Speed (WS)
///
/// returns: f64 Heading
///
/// # Examples
///
/// ```
/// use aviation_calculator::navigation::*;
///
/// let heading = heading(90.0, 110.0, 180.0, 12.5);
/// ```
pub fn heading(dc: f64, tas: f64, wd: f64, ws: f64) -> f64 {
    round(dc + wind_correction_angle(tas, ws, wd - dc), 2)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn calculate_ground_speed_no_wind() {
        let result = ground_speed(70.0, 100.0, 0.0, 0.0);
        assert_eq!(result, 100.0);
    }

    #[test]
    fn calculate_ground_speed_headwind() {
        let result = ground_speed(0.0, 100.0, 0.0, 30.0);
        assert_eq!(result, 70.0);
    }

    #[test]
    fn calculate_ground_speed_tailwind() {
        let result = ground_speed(0.0, 100.0, 180.0, 20.0);
        assert_eq!(result, 120.0);
    }

    #[test]
    fn calculate_ground_speed_crosswind() {
        let result = ground_speed(180.0, 100.0, 90.0, 10.0);
        assert_eq!(result, 99.5);
    }

    #[test]
    fn calculate_ground_speed_precision() {
        let result = ground_speed(45.0, 90.0, 90.0, 12.0);
        assert_eq!(result, 81.11);
    }

    #[test]
    fn calculate_wca_0() {
        let result = wind_correction_angle(100.0, 20.0, 90.0);
        assert_eq!(result, 11.54);
    }

    #[test]
    fn calculate_wca_1() {
        let result = wind_correction_angle(100.0, 20.0, 450.0);
        assert_eq!(result, 11.54);
    }

    #[test]
    fn calculate_wca_2() {
        let result = wind_correction_angle(100.0, 20.0, 270.0);
        assert_eq!(result, -11.54);
    }

    #[test]
    fn calculate_wca_3() {
        let result = wind_correction_angle(90.0, 20.0, 60.0);
        assert_eq!(result, 11.1);
    }

    #[test]
    fn calculate_wca_4() {
        let result = wind_correction_angle(90.0, 0.0, 0.0);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn calculate_wca_5() {
        let result = wind_correction_angle(90.0, 0.0, 360.0);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn calculate_wca_6() {
        let result = wind_correction_angle(90.0, 0.0, 40.0);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn calculate_wca_7() {
        let result = wind_correction_angle(90.0, 40.0, 0.0);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn calculate_wca_8() {
        let result = wind_correction_angle(90.0, 40.0, 180.0);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn calculate_wca_9() {
        let result = wind_correction_angle(90.0, 20.0, -45.0);
        assert_eq!(result, -9.04);
    }

    #[test]
    fn calculate_wca_10() {
        let result = wind_correction_angle(90.0, 20.0, 100.0);
        assert_eq!(result, 12.64);
    }

    #[test]
    fn calculate_wca_11() {
        let result = wind_correction_angle(95.0, 11.2, 360.0);
        assert_eq!(result, 0.0);
    }

    #[test]
    fn calculate_wca_12() {
        let result = wind_correction_angle(95.0, 11.2, -160.0);
        assert_eq!(result, -2.31);
    }

    #[test]
    fn calculate_heading_0() {
        let result = heading(90.0, 100.0, 180.0, 20.0);
        assert_eq!(result, 101.54);
    }

    #[test]
    fn calculate_heading_1() {
        let result = heading(0.0, 100.0, 90.0, 20.0);
        assert_eq!(result, 11.54);
    }

    #[test]
    fn calculate_heading_2() {
        let result = heading(320.0, 100.0, 90.0, 23.0);
        assert_eq!(result, 330.15);
    }

    #[test]
    fn calculate_heading_3() {
        let result = heading(120.0, 90.0, 70.0, 30.0);
        assert_eq!(result, 105.21);
    }

    #[test]
    fn calculate_heading_4() {
        let result = heading(350.0, 95.0, 190.0, 10.1);
        assert_eq!(result, 347.92);
    }
}


src/meteorology.rs
use snafu::prelude::*;

// https://www.dwd.de/DE/service/lexikon/begriffe/S/Standardatmosphaere_pdf.pdf?__blob=publicationFile&v=3
const ISA_TEMPERATURE: f64 = 288.15_f64; /* K */
const ISA_PRESSURE: f64 = 1013.25_f64; /* hPa */
const TROPOSPHERIC_TEMPERATURE_LAPSE: f64 = 0.0065_f64; /* K m-1 */
const STRATOSPHERIC_TEMPERATURE_LAPSE: f64 = 0.0010_f64; /* K m-1 */
const SPECIFIC_GAS_CONSTANT: f64 = 287.058_f64;
const GRAVITATIONAL_ACCELERATION: f64 = 9.81_f64; /* m/s */
const ICAO_MINIMUM_PRESSURE_ALTITUDE: f64 = -1_000.0_f64; /* m */
const ICAO_MAXIMUM_PRESSURE_ALTITUDE: f64 = 80_000.0_f64; /* m */

#[derive(Debug, Clone, Copy)]
struct AtmosphericLevel {
    base: u32,
    lapse_rate: f64,
    base_temperature: f64,
}

const LEVELS: [&AtmosphericLevel; 4] = [
    // Troposphere
    &AtmosphericLevel {
        base: 0,
        lapse_rate: TROPOSPHERIC_TEMPERATURE_LAPSE,
        base_temperature: 15.0,
    },
    // Tropopause
    &AtmosphericLevel {
        base: 11_000,
        lapse_rate: 0.0,
        base_temperature: -56.5,
    },
    // Stratosphere
    &AtmosphericLevel {
        base: 20_000,
        lapse_rate: STRATOSPHERIC_TEMPERATURE_LAPSE,
        base_temperature: -56.5,
    },
    // Stratosphere
    &AtmosphericLevel {
        base: 32_000,
        lapse_rate: 0.0,
        base_temperature: -44.5,
    },
];

#[derive(Debug, Snafu)]
pub enum UndefinedPressureAltitudeError {
    #[snafu(display("The pressure altitude {pressure_altitude} m is below the minimum defined ({min} m) in the ICAO Standard Atmosphere"))]
    BelowMinimum { min: f64, pressure_altitude: f64 },

    #[snafu(display("The pressure altitude {pressure_altitude} m is above the maximum defined ({max} m) in the ICAO Standard Atmosphere"))]
    AboveMaximum { max: f64, pressure_altitude: f64 },
}

///
///
/// # Arguments
///
/// * `pressure_altitude`: Pressure altitude in meters
///
/// returns: Result<f64, UndefinedPressureAltitudeError> Default temperature for the given pressure altitude
///
/// # Examples
///
/// ```
/// use aviation_calculator::meteorology::*;
///
/// let temp: f64 = icao_temperature(113.7).unwrap();
/// ```
pub fn icao_temperature(pressure_altitude: f64) -> Result<f64, UndefinedPressureAltitudeError> {
    if pressure_altitude < ICAO_MINIMUM_PRESSURE_ALTITUDE {
        return Err(UndefinedPressureAltitudeError::BelowMinimum { min: ICAO_MINIMUM_PRESSURE_ALTITUDE, pressure_altitude });
    }

    if pressure_altitude > ICAO_MAXIMUM_PRESSURE_ALTITUDE {
        return Err(UndefinedPressureAltitudeError::AboveMaximum { max: ICAO_MAXIMUM_PRESSURE_ALTITUDE, pressure_altitude });
    }

    let current_level = atmospheric_level_by_geopotential_altitude(pressure_altitude);

    Ok(round(current_level.base_temperature - (pressure_altitude - current_level.base as f64) * current_level.lapse_rate, 2))
}

/// # Calculate Pressure Altitude by QNH and Field Elevation
///
/// ## Arguments
///
/// * `qnh`: QNH for the location given in hPa
/// * `field_elevation`: Field elevation given in meters
///
/// returns: f64
///
/// # Examples
///
/// ```
/// use aviation_calculator::meteorology::*;
///
/// let pressure: f64 = pressure_altitude_by_qnh(1021.0, 113.7);
/// ```
pub fn pressure_altitude_by_qnh(qnh: f64, field_elevation: f64) -> f64 {
    round(field_elevation
              + ISA_TEMPERATURE / TROPOSPHERIC_TEMPERATURE_LAPSE
        * (1.0_f64
        - (qnh / ISA_PRESSURE).powf(
        SPECIFIC_GAS_CONSTANT * TROPOSPHERIC_TEMPERATURE_LAPSE
            / GRAVITATIONAL_ACCELERATION,
    )), 2)
}

///
///
/// # Arguments
///
/// * `pressure_altitude`: Pressure altitude in meters
/// * `temperature`: Current temperature in °C
///
/// returns: Result<f64, UndefinedPressureAltitudeError>
///
/// # Examples
///
/// ```
/// use aviation_calculator::meteorology::*;
///
/// let temperature_deviation = calculate_temperature_deviation(113.0, 21.0);
/// ```
pub fn calculate_temperature_deviation(pressure_altitude: f64, temperature: f64) -> Result<f64, UndefinedPressureAltitudeError> {
    Ok(round(temperature - icao_temperature(pressure_altitude)?, 2))
}

fn atmospheric_level_by_geopotential_altitude<'a>(elevation: f64) -> &'a AtmosphericLevel {
    LEVELS.iter()
        .take_while(|level| elevation >= level.base as f64)
        .last()
        .unwrap_or(LEVELS.first().unwrap())
}

fn round(x: f64, decimals: u8) -> f64 {
    let y = 10_i32.pow(decimals.into()) as f64;
    (x * y).round() / y
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn pressure_altitude_example_1() {
        let result = pressure_altitude_by_qnh(1021.0, 113.0);
        assert_eq!(result, 48.71);
    }

    #[test]
    fn pressure_altitude_example_2() {
        let result = pressure_altitude_by_qnh(1013.25, 113.0);
        assert_eq!(result, 113.0);
    }

    #[test]
    fn pressure_altitude_example_3() {
        let result = pressure_altitude_by_qnh(1021.0, 113.0);
        assert_eq!(result, 48.71);
    }

    #[test]
    fn isa_temperature_out_or_range_negative() {
        let result = icao_temperature(-1000.01);
        assert!(result.is_err());
    }

    #[test]
    fn isa_temperature_negative_1000() {
        let result = icao_temperature(-1000.0);
        assert_eq!(result.unwrap(), 21.5_f64);
    }

    #[test]
    fn isa_temperature_0() {
        let result = icao_temperature(0.0);
        assert_eq!(result.unwrap(), 15.0_f64);
    }

    #[test]
    fn isa_temperature_113() {
        let result = icao_temperature(113.7);
        assert_eq!(result.unwrap(), 14.26_f64);
    }

    #[test]
    fn isa_temperature_1000() {
        let result = icao_temperature(1000.0);
        assert_eq!(result.unwrap(), 8.5_f64);
    }

    #[test]
    fn isa_temperature_2000() {
        let result = icao_temperature(2000.0);
        assert_eq!(result.unwrap(), 2.0_f64);
    }

    #[test]
    fn isa_temperature_3000() {
        let result = icao_temperature(3000.0);
        assert_eq!(result.unwrap(), -4.5_f64);
    }

    #[test]
    fn isa_temperature_5000() {
        let result = icao_temperature(5000.0);
        assert_eq!(result.unwrap(), -17.5_f64);
    }

    #[test]
    fn isa_temperature_7000() {
        let result = icao_temperature(7000.0);
        assert_eq!(result.unwrap(), -30.5_f64);
    }

    #[test]
    fn isa_temperature_9000() {
        let result = icao_temperature(9_000.0);
        assert_eq!(result.unwrap(), -43.5_f64);
    }

    #[test]
    fn isa_temperature_11000() {
        let result = icao_temperature(11_000.0);
        assert_eq!(result.unwrap(), -56.5_f64);
    }

    #[test]
    fn isa_temperature_13000() {
        let result = icao_temperature(13_000.0);
        assert_eq!(result.unwrap(), -56.5_f64);
    }

    #[test]
    fn isa_temperature_20000() {
        let result = icao_temperature(20_000.0);
        assert_eq!(result.unwrap(), -56.5_f64);
    }

    #[test]
    fn isa_temperature_25000() {
        let result = icao_temperature(25_000.0);
        assert_eq!(result.unwrap(), -61.5_f64);
    }

    #[test]
    fn isa_temperature_32000() {
        let result = icao_temperature(32_000.0);
        assert_eq!(result.unwrap(), -44.5_f64);
    }

    #[test]
    fn isa_temperature_80000() {
        let result = icao_temperature(80_000.0);
        assert_eq!(result.unwrap(), -44.5_f64);
    }

    #[test]
    fn isa_temperature_out_of_range_positive() {
        let result = icao_temperature(80_000.01_f64);
        assert!(result.is_err());
    }

    #[test]
    fn fsm75_3_example1() {
        let result = icao_temperature(182.88);
        assert_eq!(result.unwrap(), 13.81); // 14.0
    }

    #[test]
    fn fsm75_3_example2() {
        let result = icao_temperature(609.6);
        assert_eq!(result.unwrap(), 11.04); // 11.0
    }

    #[test]
    fn fsm75_3_example3() {
        let result = icao_temperature(350.52);
        assert_eq!(result.unwrap(), 12.72); // 13.0
    }

    #[test]
    fn temperature_deviation_msl() {
        let result = calculate_temperature_deviation(0.0, 15.0);
        assert_eq!(result.unwrap(), 0.0);
    }

    #[test]
    fn temperature_deviation_positive() {
        let result = calculate_temperature_deviation(0.0, 16.0);
        assert_eq!(result.unwrap(), 1.0);
    }

    #[test]
    fn temperature_deviation_negative() {
        let result = calculate_temperature_deviation(0.0, 14.0);
        assert_eq!(result.unwrap(), -1.0);
    }

    #[test]
    fn temperature_deviation_below_msl() {
        let result = calculate_temperature_deviation(-200.0, 15.0);
        assert_eq!(result.unwrap(), -1.3);
    }

    #[test]
    fn temperature_deviation_above_msl() {
        let result = calculate_temperature_deviation(200.0, 15.0);
        assert_eq!(result.unwrap(), 1.3);
    }
}
